torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -537,6 +620,58 @@ void Reducer::initialize_buckets(
 
         // Allocate bucket contents tensor.
         replica.contents = at::empty({static_cast<long>(offset)}, options);
+
+        // Note:  "Gradient Layout Contract"
+        //
+        // Here, create views into the contents tensor for each variable's grad.
+        // Views serve as entry points to copy_ each grad's data in/out of the
+        // flat contents tensor.
+        //
+        // Gradients may have dense memory but non-row-major-contiguous strides
+        // (e.g. channels_last or channels_last_3d). For coalesced accesses during
+        // copy_s, it's beneficial for each view's layout to match its grad's layout.
+        //
+        // Specifically, we expect torch/csrc/autograd/AccumulateGrad.h produces grads
+        // that obey there "Gradient Layout Contract":
+        //   (1) if variable.is_non_overlapping_and_dense(), the stashed grad's
+        //       strides match variable.
+        //   (2) else, stashed grad is rowmajor contiguous.
+        // and create views to match.
+        //
+        // If AccumulateGrad breaks the contract, and produces a grad with an
+        // unexpected layout, performance will degrade due to poor memory access
+        // patterns when copy_ing grad data in and out of its bucket view.
```

Comment:
curious, why copy_ing instead of copying?

Reply:
Do you mean clone()?  clone() would replace the param, which would also wipe out the AccumulateGrad hook.

Reply:
do you mean clone()?  We shouldn't make variable.grad a new tensor at a new location if we don't have to.  If users reference param.grad before backward, they're accustomed to that reference remaining valid after backward under certain circumstances.  See [the docstring i added](https://github.com/pytorch/pytorch/pull/40358/files#diff-d95cdc9aa4b6a8f77487f03c988006faR62-R79) for details.  The layout behavior the docstring describes is new, but the in-place vs out-of-place behavior for grad accumulation is identical to current master.  @albanD told me that my PR should avoid changing existing in-place vs out-of-place accumulation behavior to improve BC.